### PR TITLE
[indylambda] Relieve LambdaMetafactory of boxing duties [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
`LambdaMetafactory` generates code to perform a limited number of type
adaptations when delegating from its implementation of the functional
interface method to the lambda target method.

These adaptations are: numeric widening, casting, boxing and unboxing.

However, the semantics of unboxing numerics in Java differs to Scala: they
treat `UNBOX(null)` as cause to rause a `NullPointerException`, Scala (in
`BoxesRuntime.unboxTo{Byte,Short,...}`) reinterprets the null as zero.

Furthermore, Java has no idea how to adapt between a value class and its
wrapped type, nor from a void return to `BoxedUnit`.

This commit detects when the lambda target method would require such
adaptation. If it does, an extra method, `$anonfun$1$adapted` is created to
perform the adaptation, and this is used as the target of the lambda.

This obviates the use of `JProcedureN` for `Unit` returning lambdas, we
know use `JFunctionN` as the functional interface and bind this to an
`$adapted` method that summons the instance of `BoxedUnit` after calling
the `void` returning lambda target.

The enclosed test cases fail without boxing changes. They don't execute
with indylambda enabled under regular partest runs yet, you need to add
scala-java8-compat to scala-library and pass the SCALAC_OPTS to partest
manually to try this out, as described in
https://github.com/scala/scala/pull/4463. Once we enable indylambda by
default, however, this test will exercise the code in this patch all the
time.

It is also possible to run the tests with:

```
% curl
https://oss.sonatype.org/content/repositories/releases/org/scala-lang/modules/scala-java8-t_2.11/0.4.0/scala-java8-compat_2.11-0.4.0.jar
> scala-java8-compat_2.11-0.4.0.jar
% export INDYLAMBDA="-Ydelambdafy:method -Ybackend:GenBCode -target:jvm-1.8
-classpath .:scala-java8-compat_2.11-0.4.0.jar" qscalac $INDYLAMBDA
test/files/run/indylambda-boxing/*.scala && qscala $INDYLAMBDA Test
```

### CODE REVIEW

**Comment by lrytz** [comment]
typos in the commit message: "~~rause~~|rise a `NullPointerException`", "we ~~know~~|now use `JFunctionN`"

**Comment by lrytz** [comment]
Could you remove the `JProcedureN` from https://github.com/scala/scala-java8-compat now, or are they still in use elsewhere?

**Comment by retronym** [comment]
`JProcedureN` is still used as a target type when writing a Java lambda to conform to `FunctionN[A, Unit]` without the need to explicitly return a `BoxedUnit.INSTANCE`.

File: src/compiler/scala/tools/nsc/transform/Delambdafy.scala [hunk 1]
```diff
@@ -105,6 +107,12 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
             // ... or an invokedynamic call
             super.transform(apply)
         }
+      case Template(_, _, _) =>
+        try {
+          // during this call boxingBridgeMethods will be populated from the Function case
+          val Template(parents, self, body) = super.transform(tree)
+          Template(parents, self, body ++ boxingBridgeMethods)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Delambdafy.scala:114]: I couldn't easily/lazily convince myself this deals correctly with nested templates. Could you shed some light on that?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/Delambdafy.scala:114]: The train has already been through Flatten Plains, almost at the terminal station.

```
qscala -Xshow-phases 2>&1| tail -10
    lambdalift  17  move nested functions to top level
  constructors  18  move field definitions into constructors
       flatten  19  eliminate inner classes
         mixin  20  mixin composition
       cleanup  21  platform-specific cleanups, generate reflective calls
  🚅delambdafy  22  remove lambdas
         icode  23  generate portable intermediate code
           jvm  24  generate JVM bytecode
      terminal  25  the last phase during a compilation run
```

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Delambdafy.scala:114]: Ah yes. I don't spend nearly enough time in back country.

**Comment by retronym** [comment]
@lrytz fxed the typo in the commit message

File: src/reflect/scala/reflect/internal/transform/PostErasure.scala [hunk 2]
```diff
@@ -9,7 +9,8 @@ trait PostErasure {
   object elimErasedValueType extends TypeMap {
     def apply(tp: Type) = tp match {
       case ConstantType(Constant(tp: Type)) => ConstantType(Constant(apply(tp)))
-      case ErasedValueType(_, underlying)   => underlying
+      case ErasedValueType(_, underlying)   =>
+        underlying
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/transform/PostErasure.scala:13]: `nop;`

**Comment by lrytz** [comment]
LGTM!

File: src/reflect/scala/reflect/internal/transform/PostErasure.scala (hunk 2 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/transform/PostErasure.scala:13]: I'll revert this in the followup PR.

**Pull request merged**